File: Python/Challenges/Day2.py
import sys
from typing import List, Set
from collections import Counter


def main(args: List[str]) -> None:
    """
    Application entry point
    :param args: Argument list, should contain the file to load at index 1
    """

    twos: int = 0
    threes: int = 0
    ids: List[str] = []

    # Read file
    with open(args[1], "r") as f:
        for line in f:
            ids.append(line)
            counts: Set[int] = set(Counter(line).values())
            if 2 in counts:
                twos += 1
            if 3 in counts:
                threes += 1

    print(f"Part one checksum: {twos * threes}")

    # Annotate tuple unpacking variables
    a: str
    b: str

    # While the list has more than one element
    while len(ids) > 1:
        # Get last element, keep the rest for the secondary loop
        first: str = ids.pop()
        # Get second element
        for second in ids:
            mismatch: bool = False
            diff: str = ""

            # Check character by character
            for a, b in zip(first, second):
                # If mismatch
                if a != b:
                    # If a mismatch was already found, break out
                    if mismatch:
                        break
                    # Else raise mismatch flag
                    mismatch = True

                # If not mismatched, add character to the diff
                else:
                    diff += a

            # If only one mismatch, the target has been found
            else:
                if mismatch:
                    print("Part two diff: " + diff)
                    return


# Only run if entry point
if __name__ == "__main__":
    main(sys.argv)

File: CSharp/Challenges/Day2.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AdventOfCode.Challenges
{
    public class Day2 : Challenge
    {
        #region Properties
        /// <summary>
        /// Day ID
        /// </summary>
        public override int ID { get; } = 2;
        #endregion

        #region Methods
        /// <summary>
        /// Challenge solver
        /// </summary>
        public override void Solve()
        {
            int twos = 0, threes = 0;
            Stack<string> ids = new Stack<string>();
            foreach (string line in GetLines())
            {
                ids.Push(line);
                HashSet<int> counts = new HashSet<int>(line.GroupBy(c => c).Select(g => g.Count()));
                if (counts.Contains(2)) { twos++; }
                if (counts.Contains(3)) { threes++; }
            }

            Print("Part one checksum: " + (twos * threes));

            while (ids.Count > 0)
            {
                string first = ids.Pop();
                foreach (string second in ids)
                {
                    bool mismatch = false;
                    StringBuilder sb = new StringBuilder(first.Length);
                    for (int i = 0; i < first.Length; i++)
                    {
                        if (first[i] != second[i])
                        {
                            if (mismatch)
                            {
                                mismatch = false;
                                break;
                            }
                            mismatch = true;
                        }
                        else
                        {
                            sb.Append(first[i]);
                        }
                    }

                    if (mismatch)
                    {
                        Print("Part two diff: " + sb);
                        return;
                    }
                }
            }
        }
        #endregion
    }
}
